File: 백준/Gold/2293. 동전 1/동전 1.py
import sys

n ,k = map(int,sys.stdin.readline().split())

coins = [int(sys.stdin.readline().rstrip()) for _ in range(n)]
dp = [ 0 for _ in range(k+1)]
dp[0] = 1

for coin in coins:
    for i in range(coin, k + 1):
        dp[i] += dp[i-coin]

print(dp[k])

File: 백준/Gold/2293. 동전 1/동전 1.java
import java.util.*;
import java.io.*;

class Main {
  public static void main(String[] args) throws IOException {
    BufferedReader br  = new BufferedReader(new InputStreamReader(System.in));
    
    String[] st = br.readLine().split(" ");

    int n = Integer.parseInt(st[0]);
    int k = Integer.parseInt(st[1]);
    
    int[] coins = new int[n];
    for (int i = 0; i < n; i++) coins[i] = Integer.parseInt(br.readLine());

    int[] dp = new int[k+1];
    dp[0] = 1;

    // 1 2 5
  
    for (int coin : coins) {
      for (int i = coin; i <= k; i++) {
        dp[i] += dp[i - coin];
      }
    }

    System.out.print(dp[k]);
  }
} 